File: 1016-subarray-sums-divisible-by-k/1016-subarray-sums-divisible-by-k.cpp
class Solution {
public:
    int subarraysDivByK(vector<int>& nums, int k) {
        // Create a hash map to store the frequency of each prefix sum mod k.
        unordered_map<int, int> prefixSumCount {{0, 1}};
        int countSubarrays = 0; // Initialize the number of subarrays divisible by k.
        int cumulativeSum = 0; // This will keep track of the cumulative sum of elements.

        // Iterate over each number in the array.
        for (int num : nums) {
            // Add current number to cumulative sum and do mod by k.
            // The double mod ensures that cumulativeSum is positive.
            int x = cumulativeSum + num;
            cumulativeSum = (x % k + k) % k;

            // If a subarray has cumulativeSum mod k equals to some previous subarray's cumulativeSum mod k,
            // then the subarray in between is divisible by k.
            countSubarrays += prefixSumCount[cumulativeSum]++;

            // The prefixSumCount[cumulativeSum]++ increases the count of the
            // current prefix sum mod k, which will be used for future subarray checks.
        }

        // Return the total count of subarrays with sum divisible by k.
        return countSubarrays;
    }
};

File: 1016-subarray-sums-divisible-by-k/1016-subarray-sums-divisible-by-k.java
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        int[] mc = new int[k];
        mc[0] =1;

        int cumulativeSum=0;
        int countSubarraySum=0;

        for(int num:nums){
            int x = cumulativeSum + num;
            cumulativeSum = (x % k+k)%k;

            countSubarraySum += mc[cumulativeSum];
            mc[cumulativeSum]++;
        }
        return countSubarraySum;
    }
}


/*
class Solution {
    public int subarraysDivByK(int[] nums, int k) {
        int n = nums.length;
        HashMap<Integer, Integer> mp = new HashMap<>();

        int cumulativeSum = 0;
        int countSubarraySum = 0;

        mp.put(0,1);

        for(int i = 0; i < n; i++){
            int x = cumulativeSum+nums[i];

            cumulativeSum = (x % k + k)% k;

            if(mp.containsKey(cumulativeSum) == true){
                countSubarraySum += mp.get(cumulativeSum);
            }

            mp.put(cumulativeSum, mp.getOrDefault(cumulativeSum,0)+1);
        }

        return countSubarraySum;
    }
}
*/